libs/sdk-core/src/swap_in_taproot/swap.rs

```diff
@@ -0,0 +1,1477 @@
+        Ok(FullTaprootSwapData {
+            swap,
+            parameters: Some(parameters),
+            ..Default::default()
+        })
+    }
+
+    async fn delete_invoice(&self, _swap: &FullTaprootSwapData, bolt11: String) -> Result<()> {
+        self.node.delete_invoice(bolt11).await?;
+        Ok(())
+    }
+
+    fn emit_swap_updated(&self, bitcoin_address: &str) -> Result<()> {
+        let swap_data = self
+            .persister
+            .get_full_taproot_swap(bitcoin_address)?
+            .ok_or_else(|| {
+                anyhow!(format!(
+                    "taproot swap address {} was not found",
+                    bitcoin_address
+                ))
+            })?;
+        self.status_changes_notifier
+            .send(swap_data)
+            .map_err(anyhow::Error::msg)?;
+        Ok(())
+    }
+
+    async fn get_payment_request(
+        &self,
+        swap: &FullTaprootSwapData,
+        current_tip: u32,
+    ) -> Result<(String, bool), GetPaymentRequestError> {
+        match self.get_payment_request_inner(swap, current_tip).await {
+            Ok(s) => return Ok(s),
+            Err(e) => match e {
+                GetPaymentRequestError::InvoiceAlreadyExists => {}
+                _ => return Err(e),
+            },
+        }
+
+        debug!("Retrying to get payment request because invoice already existed.");
+        // Retry getting the payment request once if it returned 'Invoice already exists' on the first try.
+        self.get_payment_request_inner(swap, current_tip).await
+    }
+
+    async fn get_payment_request_inner(
+        &self,
+        swap: &FullTaprootSwapData,
+        current_tip: u32,
+    ) -> Result<(String, bool), GetPaymentRequestError> {
+        let maybe_bolt11_result = self
+            .node
+            .fetch_bolt11(swap.swap.payment_hash.clone())
+            .await?;
+        let initial_fee_params_valid = swap
+            .swap
+            .accepted_opening_fee_params
+            .valid_for(MIN_OPENING_FEE_PARAMS_VALIDITY_SECONDS)?;
+        let opening_fee_params = match initial_fee_params_valid {
+            true => Some(swap.swap.accepted_opening_fee_params.clone()),
+            false => None,
+        };
+
+        // If a payment was requested before, the could be an existing invoice.
+        // Validate the existing invoice, it may need to be recreated.
+        if let Some(bolt11_result) = maybe_bolt11_result {
+            let maybe_bolt11 = self
+                .check_existing_payment_request(swap, bolt11_result)
+                .await?;
+            if let Some(bolt11) = maybe_bolt11 {
+                return Ok((bolt11, false));
+            }
+        };
+
+        let amount_msat = swap.confirmed_unspent_amount_msat();
+        let blocks_left = swap.blocks_left(current_tip).ok_or(anyhow!(
+            "cannot create payment request for unconfirmed swap"
+        ))?;
+        let blocks_left = match blocks_left {
+            b if b <= 0 => 1,
```

Comment:
Can you clarify @JssDWt. It looks like if we reach the expiry block (or even go past it) we can still create a invoice and complete the swap. Is that right?